File: r/tests/testthat/test_summarize_identification.R
context("Testing identification summarization functionality")

test_that("summarize_identification returns named list with expected fields", {
  ooi <- list(c(1L, 3L), c(2L, 3L))
  sizes <- c(50L, 40L)
  out <- summarize_identification(ooi, sizes, r = 2L)
  expect_type(out, "list")
  expect_setequal(names(out), c(
    "largest_super_cohort_size",
    "largest_super_cohort_share",
    "min_cohort_size_in_largest_super",
    "num_outcomes_in_largest_super_cohort",
    "num_o3_iterations"
  ))
  expect_true(out$largest_super_cohort_size >= 0)
  expect_true(out$min_cohort_size_in_largest_super >= 0)
  expect_gte(out$largest_super_cohort_share, 0)
  expect_lte(out$largest_super_cohort_share, 1)
  # With r=2 and these cohorts, no merges occur; only the initial state is present
  expect_equal(out$num_o3_iterations, 1)
  expect_equal(out$largest_super_cohort_size, 50)
  expect_equal(out$min_cohort_size_in_largest_super, 50)
  expect_equal(out$largest_super_cohort_share, 50/90, tolerance = 1e-12)
  # Largest super cohort is the larger single cohort {1,3} (1-based); union outcomes count is 2
  expect_equal(out$num_outcomes_in_largest_super_cohort, 2)
})

test_that("summarize_identification returns list of named lists for many panels", {
  ooi_list <- list(list(c(1L), c(2L, 3L)))
  size_list <- list(c(10L, 25L))
  out <- summarize_identification(ooi_list, size_list, r = 2L)
  expect_true(is.list(out))
  expect_equal(length(out), 1)
  el <- out[[1]]
  expect_setequal(names(el), c(
    "largest_super_cohort_size",
    "largest_super_cohort_share",
    "min_cohort_size_in_largest_super",
    "num_outcomes_in_largest_super_cohort",
    "num_o3_iterations"
  ))
  # For this case, no merges occur; only initial state is present
  expect_equal(el$num_o3_iterations, 1)
  expect_equal(el$largest_super_cohort_size, 25)
  expect_equal(el$min_cohort_size_in_largest_super, 25)
  expect_equal(el$largest_super_cohort_share, 25/35, tolerance = 1e-12)
  # Largest super cohort is the larger single cohort {2,3} (1-based); union outcomes count is 2
  expect_equal(el$num_outcomes_in_largest_super_cohort, 2)
})

test_that("aligned_factors_identified returns TRUE for staircase pattern", {
  observed_outcome_indices <- list(
    c(1, 2, 3),
    c(2, 3, 4),
    c(3, 4, 5)
  )
  r <- 2L
  expect_true(aligned_factors_identified(observed_outcome_indices, r))
})

test_that("aligned_factors_identified returns TRUE for non-contiguous pattern", {
  observed_outcome_indices <- list(
    c(1, 2, 3), # Cohort 1
    c(2, 3, 4), # Cohort 2
    c(1, 4, 5)  # Cohort 3
  )
  r <- 2L
  expect_true(aligned_factors_identified(observed_outcome_indices, r))
})

test_that("aligned_factors_identified returns FALSE for unconnected cohorts", {
  observed_outcome_indices <- list(
    c(1, 2), # Cohort 1
    c(2, 3, 4), # Cohort 2
    c(3, 4, 5)     # This cohort has no overlap with the others
  )
  r <- 2L
  expect_false(aligned_factors_identified(observed_outcome_indices, r))
})

test_that("aligned_factors_identified returns FALSE when no merges occur", {
  observed_outcome_indices <- list(
    c(1, 2),
    c(2, 3),
    c(3, 4, 5)
  )
  r <- 2L
  expect_false(aligned_factors_identified(observed_outcome_indices, r))
})

test_that("count_outcomes_with_rank_overlap_per_cohort matches expected counts", {
  ooi <- list(
    c(1L, 2L, 3L),
    c(2L, 3L, 4L),
    c(5L)
  )
  counts <- count_outcomes_with_rank_overlap_per_cohort(ooi, rank = 2L)
  expect_equal(as.integer(counts), c(4L, 4L, 1L))

  counts_all <- count_outcomes_with_rank_overlap_per_cohort(ooi, rank = 0L)
  expect_equal(as.integer(counts_all), c(5L, 5L, 5L))
})

test_that("get_masked_observed_outcome_indices masks per cohort as expected", {
  ooi <- list(c(1L, 3L), c(2L, 3L))
  # Drop outcome 3 from cohort 1 only
  mask <- list(`1` = c(3L))
  res <- get_masked_observed_outcome_indices(ooi, mask)
  expect_equal(res[[1]], c(1L))
  expect_equal(res[[2]], c(2L, 3L))

  # NULL mask returns unchanged
  res2 <- get_masked_observed_outcome_indices(ooi, NULL)
  expect_identical(res2, ooi)

  # Empty mask returns unchanged
  res3 <- get_masked_observed_outcome_indices(ooi, list())
  expect_identical(res3, ooi)
})

File: core/tests/test_summarize_identification.cpp
#include <gtest/gtest.h>
#include <vector>
#include <set>
#include <armadillo>
#include "summarize_identification.h"

// Helpers to build simple ObservedOutcomeIndices inputs
static std::vector<arma::uvec> make_ooi(std::initializer_list<std::initializer_list<arma::uword>> lists) {
    std::vector<arma::uvec> out;
    out.reserve(lists.size());
    for (const auto &lst : lists) {
        arma::uvec v(lst.size());
        arma::uword i = 0;
        for (arma::uword x : lst) v(i++) = x;
        out.push_back(v);
    }
    return out;
}

TEST(SummarizeIdentificationTest, SinglePanel_NoMerges_FinalIteration) {
    // Equivalent to R test: sizes <- c(50L, 40L), ooi <- list(c(1L,3L), c(2L,3L)), r=2
    // Note: C++ is 0-indexed for outcome indices; R used 1-based.
    std::vector<arma::uvec> ooi = make_ooi({{0,2}, {1,2}});
    arma::uvec sizes = {50u, 40u};
    const std::size_t r = 2;

    apm::IdentificationSummary out = apm::summarize_identification(ooi, sizes, r);

    EXPECT_GE(out.largest_super_cohort_size, static_cast<std::size_t>(0));
    EXPECT_GE(out.min_cohort_size_in_largest_super, static_cast<std::size_t>(0));
    EXPECT_GE(out.largest_super_cohort_share, 0.0);
    EXPECT_LE(out.largest_super_cohort_share, 1.0);
    EXPECT_GE(out.num_outcomes_in_largest_super_cohort, static_cast<std::size_t>(0));
    // With r=2 and these cohorts, no merges; only initial state is present
    EXPECT_EQ(out.num_o3_iterations, static_cast<std::size_t>(1));
    EXPECT_EQ(out.largest_super_cohort_size, static_cast<std::size_t>(50));
    EXPECT_EQ(out.min_cohort_size_in_largest_super, static_cast<std::size_t>(50));
    EXPECT_NEAR(out.largest_super_cohort_share, 50.0/90.0, 1e-12);
    // Largest super cohort is the larger single cohort {0,2}; union outcomes count is 2
    EXPECT_EQ(out.num_outcomes_in_largest_super_cohort, static_cast<std::size_t>(2));
    EXPECT_DOUBLE_EQ(out.total_outcome_weight_in_largest_super_cohort, 2.0);
    EXPECT_NEAR(out.share_outcomes_in_largest_super_cohort, 2.0/3.0, 1e-12);
    EXPECT_NEAR(out.share_outcome_weight_in_largest_super_cohort, 2.0/3.0, 1e-12);
}

TEST(SummarizeIdentificationTest, ManyPanels_ListInputs_NoMerges) {
    // R analog: ooi_list <- list(list(c(1L), c(2L,3L))); size_list <- list(c(10L,25L)); r=2
    std::vector<std::vector<arma::uvec>> ooi_list = { make_ooi({{0}, {1,2}}) };
    std::vector<arma::uvec> size_list = { arma::uvec{10u, 25u} };
    const std::size_t r = 2;

    std::vector<apm::IdentificationSummary> res = apm::summarize_identification(ooi_list, size_list, r);
    ASSERT_EQ(res.size(), static_cast<std::size_t>(1));
    const auto &el = res[0];
    EXPECT_EQ(el.num_o3_iterations, static_cast<std::size_t>(1));
    EXPECT_EQ(el.largest_super_cohort_size, static_cast<std::size_t>(25));
    EXPECT_EQ(el.min_cohort_size_in_largest_super, static_cast<std::size_t>(25));
    EXPECT_NEAR(el.largest_super_cohort_share, 25.0/35.0, 1e-12);
    // Largest super cohort is the larger single cohort {1,2}; union outcomes count is 2
    EXPECT_EQ(el.num_outcomes_in_largest_super_cohort, static_cast<std::size_t>(2));
    EXPECT_DOUBLE_EQ(el.total_outcome_weight_in_largest_super_cohort, 2.0);
    EXPECT_NEAR(el.share_outcomes_in_largest_super_cohort, 2.0/3.0, 1e-12);
    EXPECT_NEAR(el.share_outcome_weight_in_largest_super_cohort, 2.0/3.0, 1e-12);
}

// Move aligned_factors_identified tests here from test_apm_core.cpp
TEST(SummarizeIdentificationTest, AlignedFactorsIdentified_StaircasePattern) {
    std::vector<arma::uvec> ooi = make_ooi({{0,1,2}, {1,2,3}, {2,3,4}});
    unsigned int r = 2;
    ASSERT_TRUE(apm::aligned_factors_identified(ooi, r));
}

TEST(SummarizeIdentificationTest, AlignedFactorsIdentified_NonContiguous_TwoSteps) {
    std::vector<arma::uvec> ooi = make_ooi({{0,1,2}, {1,2,3}, {0,3,4}});
    unsigned int r = 2;
    ASSERT_TRUE(apm::aligned_factors_identified(ooi, r));
}

TEST(SummarizeIdentificationTest, AlignedFactorsNotIdentified_OneIteration) {
    std::vector<arma::uvec> ooi = make_ooi({{0,1}, {1,2,3}, {2,3,4}});
    unsigned int r = 2;
    ASSERT_FALSE(apm::aligned_factors_identified(ooi, r));
}

TEST(SummarizeIdentificationTest, AlignedFactorsNotIdentified_NoMerges) {
    std::vector<arma::uvec> ooi = make_ooi({{0,1}, {1,2}, {2,3,4}});
    unsigned int r = 2;
    ASSERT_FALSE(apm::aligned_factors_identified(ooi, r));
}

TEST(SummarizeIdentificationTest, CountOutcomesWithRankOverlap_BasicScenario) {
    std::vector<arma::uvec> ooi = make_ooi({{0,1,2}, {1,2,3}, {4}});
    const std::size_t rank = 2;

    arma::vec counts = apm::count_outcomes_with_rank_overlap_per_cohort(ooi, rank);
    ASSERT_EQ(counts.n_elem, static_cast<arma::uword>(3));
    EXPECT_DOUBLE_EQ(counts(0), 4.0);
    EXPECT_DOUBLE_EQ(counts(1), 4.0);
    // The focal cohort always contributes its own outcomes even when it has
    // fewer than `rank` observed outcomes.
    EXPECT_DOUBLE_EQ(counts(2), 1.0);
}

TEST(SummarizeIdentificationTest, CountOutcomesWithRankOverlap_RankZeroIncludesAll) {
    std::vector<arma::uvec> ooi = make_ooi({{0}, {1,2}, {2,3}});
    const std::size_t rank = 0;

    arma::vec counts = apm::count_outcomes_with_rank_overlap_per_cohort(ooi, rank);
    ASSERT_EQ(counts.n_elem, static_cast<arma::uword>(3));
    EXPECT_DOUBLE_EQ(counts(0), 4.0);
    EXPECT_DOUBLE_EQ(counts(1), 4.0);
    EXPECT_DOUBLE_EQ(counts(2), 4.0);
}

TEST(SummarizeIdentificationTest, WeightedSummariesAndCounts) {
    std::vector<arma::uvec> ooi = make_ooi({{0,2}, {1,2}});
    arma::uvec sizes = {50u, 40u};
    arma::vec weights = {1.0, 5.0, 10.0};
    const std::size_t r = 2;

    apm::IdentificationSummary out = apm::summarize_identification(ooi, sizes, r, -1, weights);
    EXPECT_EQ(out.num_outcomes_in_largest_super_cohort, static_cast<std::size_t>(2));
    EXPECT_DOUBLE_EQ(out.total_outcome_weight_in_largest_super_cohort, 11.0);
    EXPECT_NEAR(out.share_outcomes_in_largest_super_cohort, 2.0/3.0, 1e-12);
    EXPECT_NEAR(out.share_outcome_weight_in_largest_super_cohort, 11.0/16.0, 1e-12);

    arma::vec weighted_counts = apm::count_outcomes_with_rank_overlap_per_cohort(ooi, 1, weights);
    ASSERT_EQ(weighted_counts.n_elem, static_cast<arma::uword>(2));
    EXPECT_DOUBLE_EQ(weighted_counts(0), 16.0);
    EXPECT_DOUBLE_EQ(weighted_counts(1), 16.0);
}